pkg/gadgets/ebpftop/piditer/iter.go

```diff
@@ -0,0 +1,191 @@
+
+type pidIter struct {
+	objs piditerObjects
+	iter *link.Iter
+}
+
+type PidIterEntry struct {
+	ProgID uint32
+	Pid    uint32
+	Comm   string
+}
+
+var (
+	addrLock        sync.Mutex
+	bpfProgFopsAddr uint64
+	triedGetAddr    bool
+
+	iterEntrySize = int(unsafe.Sizeof(piditerPidIterEntry{}))
+)
+
+// getBpfProgFopsAddr reads the address of bpf_prog_fops from /proc/kallsyms. It
+// only does this on first call and caches the address.
+func getBpfProgFopsAddr() uint64 {
+	addrLock.Lock()
+	defer addrLock.Unlock()
+
+	if triedGetAddr {
+		return bpfProgFopsAddr
+	}
+
+	triedGetAddr = true
+
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return 0
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		if fields[2] == "bpf_prog_fops" {
+			addr, err := strconv.ParseUint(fields[0], 16, 64)
+			if err != nil {
+				return 0
+			}
+
+			// cache addr
+			bpfProgFopsAddr = addr
+			return bpfProgFopsAddr
+		}
+	}
+	return bpfProgFopsAddr
+}
+
+func NewPidIter() (*pidIter, error) {
+	spec, err := loadPiditer()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	consts := map[string]interface{}{
+		"bpf_prog_fops_addr": getBpfProgFopsAddr(),
+	}
+
+	if err := spec.RewriteConstants(consts); err != nil {
+		return nil, fmt.Errorf("error RewriteConstants: %w", err)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	opts := ebpf.CollectionOptions{}
+
+	p := &pidIter{}
+	if err := spec.LoadAndAssign(&p.objs, &opts); err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	iter, err := link.AttachIter(link.IterOptions{
+		Program: p.objs.GadgetEbpftop,
+		Map:     nil,
+	})
+	if err != nil {
+		p.objs.Close()
+		return nil, fmt.Errorf("failed to attach iter: %w", err)
+	}
+	p.iter = iter
+
+	return p, nil
+}
+
+func cstrlen(n []uint8) int {
+	for i := 0; i < len(n); i++ {
+		if n[i] == 0 {
+			return i
+		}
+	}
+	return len(n)
+}
+
+// DumpPids returns an array of PidIterEntry containing information
+// on which pid (and comm) has an open fd to which eBPF Program ID.
+func (p *pidIter) DumpPids() ([]*PidIterEntry, error) {
+	rc, err := p.iter.Open()
+	if err != nil {
+		return nil, fmt.Errorf("failed to open iter: %w", err)
+	}
+	defer rc.Close()
+
+	res := make([]*PidIterEntry, 0)
+
+	buf := make([]byte, 4096/iterEntrySize*iterEntrySize)
+	for {
+		n, err := io.ReadFull(rc, buf)
+		if err != nil {
+			if errors.Is(err, unix.EAGAIN) {
+				continue
+			}
+		}
+		if n == 0 {
+			break
+		}
+		if n%iterEntrySize != 0 {
+			return nil, fmt.Errorf("invalid format: %d", n)
+		}
+		for i := 0; i < n/iterEntrySize; i++ {
+			entry := (*piditerPidIterEntry)(unsafe.Pointer(&buf[i*iterEntrySize]))
+			res = append(res, &PidIterEntry{
+				ProgID: entry.Id,
+				Pid:    entry.Pid,
+				Comm:   string(entry.Comm[:cstrlen(entry.Comm[:])]),
```

Comment:
In various gadgets, we didn't take special care for checking the NULL-terminated string in the `comm[16]` field:
```
$ git grep -n 'GoString.*comm'
pkg/gadgets/audit-seccomp/tracer/tracer.go:165:                 Comm:      C.GoString(&eventC.comm[0]),
pkg/gadgets/execsnoop/tracer/core/tracer.go:162:                        Comm:      C.GoString(&eventC.comm[0]),
pkg/gadgets/filetop/tracer/tracer.go:187:                       Comm:       C.GoString(&fileStat.comm[0]),
pkg/gadgets/mountsnoop/tracer/core/tracer.go:178:                       Comm:      C.GoString(&eventC.comm[0]),
pkg/gadgets/oomkill/tracer/tracer.go:154:                       TriggeredComm: C.GoString(&eventC.fcomm[0]),
pkg/gadgets/oomkill/tracer/tracer.go:156:                       KilledComm:    C.GoString(&eventC.tcomm[0]),
pkg/gadgets/opensnoop/tracer/core/tracer.go:198:                        Comm:      C.GoString(&eventC.comm[0]),
pkg/gadgets/sigsnoop/tracer/core/tracer.go:234:                 Comm:      C.GoString(&eventC.comm[0]),
```

https://github.com/kinvolk/inspektor-gadget/blob/04c2a3664d830cfc39c294f9a37b1b73159de008/pkg/gadgets/oomkill/tracer/tracer.go#L154-L156

Last time I checked, the kernel was not allowed the `comm[16]` field to be non NULL-terminated. But I agree this is fragile, I don't know if future kernel could change that.

In any case, I think it should be consistent among various gadgets. So either:
* use `C.GoString(&entry.comm[0])` and remove `cstrlen`, or
* write a helper function to convert the `comm[16]` to a Go string in some shared package and use it in all gadgets

cc @mauriciovasquezbernal Any opinion?

Reply:
Good question. I think it's a potential bug and we should handle the case when comm is not null terminated. I'd suggest to use `C.GoString(&entry.comm[0])` on this PR and open an issue to track and fix it in all gadgets in a single pass. 

> write a helper function to convert the comm[16] to a Go string in some shared package and use it in all gadgets

It seems like the easiest / safest way to go.